Adding test of Gwen's activity

gkirschke opened the pull request:
Example files from class; will update later

gkirschke committed:
add new files

gkirschke committed:
updated files, not yet complete

chaedri commented on activities/Gwen.json:
```diff
@@ -0,0 +1,24 @@
+      "analyses": "Gwen.py",
+      "title": "Selecting locations for pollinator gardens that minimize overall travel time between floral resources",
+      "author": "Gwen Kirschke",
+      "instructions": "Add garden locations and program will calculate minimum distance between gardens"
```
So cool!!!

petrasovaa replied:
How do you define minimum distance between gardens?

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+    coeff = gcore.parse_command(
+        "r.regression.line", mapx=after, mapy=before, flags="g", env=env
+    )
+    grast.mapcalc(
```
```suggestion
    gs.mapcalc(
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+):
+    diff_thr = "diff_thr_" + str(uuid.uuid4()).replace("-", "")
+    diff_thr_clump = "diff_thr_clump_" + str(uuid.uuid4()).replace("-", "")
+    coeff = gcore.parse_command(
```
```suggestion
    coeff = gs.parse_command(
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+    )
+    try:
+        if add:
+            grast.mapcalc(
```
```suggestion
            gs.mapcalc(
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+                env=env,
+            )
+        else:
+            grast.mapcalc(
```
```suggestion
            gs.mapcalc(
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+                env=env,
+            )
+
+        gcore.run_command("r.clump", input=diff_thr, output=diff_thr_clump, env=env)
```
```suggestion
        gs.run_command("r.clump", input=diff_thr, output=diff_thr_clump, env=env)
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+
+        gcore.run_command("r.clump", input=diff_thr, output=diff_thr_clump, env=env)
+        stats = (
+            gcore.read_command(
```
```suggestion
            gs.read_command(
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+                    cats.append(cat)
+            if cats:
+                rules = ["{c}:{c}:1".format(c=c) for c in cats]
+                gcore.write_command(
```
```suggestion
                gs.write_command(
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+                    stdin="\n".join(rules),
+                    env=env,
+                )
+                gcore.run_command(
```
```suggestion
                gs.run_command(
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+                    env=env,
+                )
+            else:
+                gcore.warning("No change found!")
```
```suggestion
                gs.warning("No change found!")
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+                )
+            else:
+                gcore.warning("No change found!")
+                gcore.run_command("v.edit", map=change, tool="create", env=env)
```
```suggestion
                gs.run_command("v.edit", map=change, tool="create", env=env)
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+                gcore.warning("No change found!")
+                gcore.run_command("v.edit", map=change, tool="create", env=env)
+        else:
+            gcore.warning("No change found!")
```
```suggestion
            gs.warning("No change found!")
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+                gcore.run_command("v.edit", map=change, tool="create", env=env)
+        else:
+            gcore.warning("No change found!")
+            gcore.run_command("v.edit", map=change, tool="create", env=env)
```
```suggestion
            gs.run_command("v.edit", map=change, tool="create", env=env)
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+            gcore.warning("No change found!")
+            gcore.run_command("v.edit", map=change, tool="create", env=env)
+
+        gcore.run_command(
```
```suggestion
        gs.run_command(
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+            env=env,
+        )
+    except:
+        gcore.run_command(
```
```suggestion
        gs.run_command(
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+import grass.script as gs
+
+import numpy as np
+
```
Found this in tangible_utils and you use it later.

```suggestion
def get_environment(**kwargs):
    """!Returns environment for running modules.
    All modules for which region is important should
    pass this environment into run_command or similar.

    @param tmp_regions a list of temporary regions
    @param kwargs arguments for g.region

    @return environment as a dictionary
    """
    env = os.environ.copy()
    env["GRASS_OVERWRITE"] = "1"
    env["GRASS_VERBOSE"] = "0"
    env["GRASS_MESSAGE_FORMAT"] = "standard"
    region3d = False
    if "raster_3d" in kwargs:
        region3d = True
    env["GRASS_REGION"] = gs.region_env(region3d=region3d, **kwargs)
    return env

```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+        )
+
+def run_distance(real_elev, scanned_elev, env, **kwargs):
+    env2 = get_environment(raster=scanned_elev, n='n-100', s='s+100', e='e-100', w='w+100') # not sure what shifting this does?   
```
This shifting seems to be making the computational region smaller (removing a buffer of 100 on all sides). @petrasovaa might have an idea why this is necessary.

petrasovaa replied:
That's for practical reasons when scanning, the edges of the model can get messy. But you can remove that.

chaedri commented on activities/Gwen.json:
```diff
@@ -0,0 +1,24 @@
+        "smooth": 10,
+        "zexag": 2,
+        "numscans": 1,
+        "interpolate": true
```
```suggestion
        "interpolate": true,
        "calibrate":true
```

chaedri commented on activities/Gwen.py:
```diff
@@ -0,0 +1,160 @@
+                              change='change', height_threshold=[15, 100], cells_threshold=[5, 100], add=True, max_detected=13, debug=True, env=env2)
+    # dict of new points
+    points = {}
+    data = gs.read_command('v.out.ascii', input='trail_points', type='point', format='point', env=env).strip() # where does trail_points come from?
```
`trail_points` is the A and B points from that trail exercise that already exist. A little future down in the trails code, you can see the points get extracted from the change vector: https://github.com/tangible-landscape/tangible-landscape-applications/blob/4a63c3ae673139fc29ba3a92ce9acbe51bedbcd5/planting/workflows/trails.py#L41

In other words, delete this block of code and replace it with something like:

```
     points = {}
    gs.run_command('v.edit', tool='create', map=trail, env=env)
    # detected points
    points_raw = gscript.read_command('v.out.ascii', input='change',
                                      type='point', format='point').strip().split()
    for point in points_raw:
        point = point.split('|')
        point = (float(point[0]), float(point[1]))
        points[i] = point
        i += 1
```

gkirschke committed:
formatting, removing residual code from example

petrasovaa reviewed (commented):
Why don't you use the advanced_example.py, that was meant as a starting point for tasks that involve marker detection.

petrasovaa commented:
To include a dashboard, check out https://github.com/tangible-landscape/grass-tangible-landscape/blob/master/activities_config_documentation.md#additional-widgets

gkirschke committed:
changed code used for calculations; added dashboard

petrasovaa commented:
Looks like your json is invalid and the python file has a syntax error as well.

gkirschke committed:
commented out dashboard code, fixed flake errors

gkirschke commented:
I'm not sure what the issue is with the json file; I thought that what I'd
added for the dashboard was the problem, but I removed that and it still
failed. I ran the python file through black locally; what else can I do to
identify or fix the remaining black failures?

--


Gwen Kirschke

Ph.D. Student, Geospatial Analytics | Center for Geospatial Analytics
| Youngsteadt
Urban Ecology Lab


Jordan Hall 5117 | Campus Box 7106

North Carolina State University

2800 Faucette Dr.

Raleigh, NC 27695 USA

***@***.*** | geospatial.ncsu.edu



"All electronic mail messages in connection with State business which are
sent to or received by this account are subject to the NC Public Records
Law and may be disclosed to third parties.”



On Thu, Apr 11, 2024 at 12:09 PM Anna Petrasova ***@***.***>
wrote:

> Looks like your json is invalid and the python file has a syntax error as
> well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ncsu-geoforall-lab/gis714-2024-tangible-landscape/pull/3#issuecomment-2050042644>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BEELWQW4X57PVYARLQSPES3Y42YUFAVCNFSM6AAAAABEJZC6OSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANJQGA2DENRUGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> <ncsu-geoforall-lab/gis714-2024-tangible-landscape/pull/3/c2050042644@
> github.com>
>

petrasovaa commented:
Click at Code Quality / Black details:
```
+++ /home/runner/work/gis714-2024-tangible-landscape/gis714-2024-tangible-landscape/activities/Gwen.py	2024-04-11 17:37:27.843997+00:00
@@ -82,13 +82,11 @@
     env["GRASS_OVERWRITE"] = "1"
     elevation = "elev_lid792_1m"
     elev_resampled = "elev_resampled"
     # We use resampling to get a similar resolution as with Tangible Landscape.
     gs.run_command("g.region", raster=elevation, res=4, flags="a", env=env)
-    gs.run_command(
-        "r.resamp.stats", input=elevation, output=elev_resampled, env=env
-    )
+    gs.run_command("r.resamp.stats", input=elevation, output=elev_resampled, env=env)
     # The end of the block which needs no editing.
 
     # Code specific to testing of the analytical function.
     # Create points which is the additional input needed for the process.
     points = "points"
@@ -100,12 +98,10 @@
         separator="comma",
         stdin="638[43](https://github.com/ncsu-geoforall-lab/gis714-2024-tangible-landscape/actions/runs/8650940174/job/23720512664?pr=3#step:3:45)2,220382\n638621,220607",
         env=env,
     )
     # Call the analysis.
-    run_function_with_points(
-        scanned_elev=elev_resampled, env=env, points=points
-    )
+    run_function_with_points(scanned_elev=elev_resampled, env=env, points=points)
```
and fix it accordingly.

JSON file is invalid, use an online json validator to see the problem.

petrasovaa commented on activities/Gwen.py:
```diff
@@ -0,0 +1,111 @@
+    # calculate connectivity metric
+    data = gs.read_command("r.what", map="cost_surface", points=points)
+    data_lines = data.splitlines()
+    data = data_lines.split("|")
```
From continuous integration failure:
```
Traceback (most recent call last):
  File "/home/runner/work/gis714-2024-tangible-landscape/gis714-2024-tangible-landscape/activities/Gwen.py", line 111, in <module>
    main()
  File "/home/runner/work/gis714-2024-tangible-landscape/gis714-2024-tangible-landscape/activities/Gwen.py", line 105, in main
    run_function_with_points(
  File "/home/runner/work/gis714-2024-tangible-landscape/gis714-2024-tangible-landscape/activities/Gwen.py", line 63, in run_function_with_points
    data = data_lines.split("|")
AttributeError: 'list' object has no attribute 'split'
```
Splitlines() returns list, not string, so you have to iterate through the list and call split on each line.

gkirschke committed:
fixed code for calculating connectivity metric, removed dashboard from json

petrasovaa commented on activities/Gwen.py:
```diff
@@ -0,0 +1,89 @@
+        )
+
+        # generate random point for focal garden
+        gs.run_command("v.random", output="focal_garden", npoints=1, seed=3)
+
+        # calculate connectivity between focal garden and points from pins
+        gs.run_command(
+            "r.cost",
+            input="cfactorgrow_1m",
+            start_points="focal_garden",
+            stop_points="points",
+            output="cost_surface",
+        )
+        data = gs.read_command("r.what", map="cost_surface", points="points")
+        data_lines = data.splitlines()
+        connectivity = 0
+        for line in data_lines:
+            line = line.split("|")
+            line = float(line[3])
+            connectivity += 1 / line
+
```
This entire block needs to be outside of the `if not points` condition

petrasovaa commented on activities/Gwen.py:
```diff
@@ -0,0 +1,89 @@
+            "r.cost",
+            input="cfactorgrow_1m",
+            start_points="focal_garden",
+            stop_points="points",
```
You want `points` variable here.
```suggestion
            stop_points=points,
```

gkirschke commented:
I still don't have a way to display the 'connectivity' value as my output on TL; when I have tried to set up a dashboard before I wasn't getting it right. Does using print() in this interface do anything (even to display the value on the computer for the purposes of this class project)? Otherwise, could someone help me solve the problems I was having with the dashboard?

There seem to be two things happening in the check failures, and I'm not sure how to solve either of them. The render failure can't find the cost surface map that I make in the run_ function, which I thought was available to it. Do I need a placeholder map (in main()?) with that name to display until the run_ function is called? The black issues seem to be with the code in main() that sets the region and makes the test points file. Since I didn't write this code, I'm not sure how I should modify it so that it still runs?

petrasovaa commented:
> There seem to be two things happening in the check failures, and I'm not sure how to solve either of them. The render failure can't find the cost surface map that I make in the run_ function, which I thought was available to it. Do I need a placeholder map (in main()?) with that name to display until the run_ function is called? The black issues seem to be with the code in main() that sets the region and makes the test points file. Since I didn't write this code, I'm not sure how I should modify it so that it still runs?

See my comments above. Regarding black, just follow what the check suggests (- means what was there before and + means what black wants you to change it to). See also https://github.com/ncsu-geoforall-lab/gis714-2024-tangible-landscape/blob/master/activities/advanced_example.py, the code is aligned to what black suggests.

petrasovaa commented:
> I still don't have a way to display the 'connectivity' value as my output on TL; when I have tried to set up a dashboard before I wasn't getting it right. Does using print() in this interface do anything (even to display the value on the computer for the purposes of this class project)? Otherwise, could someone help me solve the problems I was having with the dashboard?

You can put the dashboard code you wrote previously in there, but comment it out, because it won't work with the checks, but I can make it run on Tangible Landscape.

gkirschke committed:
moved calculation outside of condition, formatting

petrasovaa commented on activities/Gwen.py:
```diff
@@ -0,0 +1,85 @@
+        )
+
+    # generate random point for focal garden
+    gs.run_command("v.random", output="focal_garden", npoints=1, seed=3)
```
You need to pass the `env` variable everywhere.

gkirschke committed:
re-added dashboard code, let overwrite on subsequent runs

gkirschke committed:
passing env argument

gkirschke committed:
removing dashboard from json file again

gkirschke commented:
> You can put the dashboard code you wrote previously in there, but comment it out, because it won't work with the checks, but I can make it run on Tangible Landscape.

Can the dashboard be in the json file? Or will that give an error too, since it won't be able to find it while it is commented out?

petrasovaa commented:
> > You can put the dashboard code you wrote previously in there, but comment it out, because it won't work with the checks, but I can make it run on Tangible Landscape.
> 
> Can the dashboard be in the json file? Or will that give an error too, since it won't be able to find it while it is commented out?

I believe you can add it to the JSON file. JSON doesn't have comments, so just leave it there.

gkirschke committed:
adding dashboard back to json

gkirschke committed:
fixed json syntax
